packages/contracts/src/utils/core_combinatorial_utils.ts

```diff
@@ -0,0 +1,809 @@
+            AssetDataScenario.ZRXFeeToken,
+        ];
+        const takerAssetDataScenario = [
+            AssetDataScenario.ERC20FiveDecimals,
+            AssetDataScenario.ERC20NonZRXEighteenDecimals,
+            AssetDataScenario.ERC721,
+            AssetDataScenario.ZRXFeeToken,
+        ];
+        const takerAssetFillAmountScenario = [TakerAssetFillAmountScenario.ExactlyRemainingFillableTakerAssetAmount];
+        const fillScenarioArrays = CoreCombinatorialUtils._getAllCombinations([
+            takerScenarios,
+            feeRecipientScenarios,
+            makerAssetAmountScenario,
+            takerAssetAmountScenario,
+            makerFeeScenario,
+            takerFeeScenario,
+            expirationTimeSecondsScenario,
+            makerAssetDataScenario,
+            takerAssetDataScenario,
+            takerAssetFillAmountScenario,
+        ]);
+
+        const fillScenarios = _.map(fillScenarioArrays, fillScenarioArray => {
+            const fillScenario: FillScenario = {
+                orderScenario: {
+                    takerScenario: fillScenarioArray[0] as TakerScenario,
+                    feeRecipientScenario: fillScenarioArray[1] as FeeRecipientAddressScenario,
+                    makerAssetAmountScenario: fillScenarioArray[2] as OrderAssetAmountScenario,
+                    takerAssetAmountScenario: fillScenarioArray[3] as OrderAssetAmountScenario,
+                    makerFeeScenario: fillScenarioArray[4] as OrderAssetAmountScenario,
+                    takerFeeScenario: fillScenarioArray[5] as OrderAssetAmountScenario,
+                    expirationTimeSecondsScenario: fillScenarioArray[6] as ExpirationTimeSecondsScenario,
+                    makerAssetDataScenario: fillScenarioArray[7] as AssetDataScenario,
+                    takerAssetDataScenario: fillScenarioArray[8] as AssetDataScenario,
+                },
+                takerAssetFillAmountScenario: fillScenarioArray[9] as TakerAssetFillAmountScenario,
+                makerStateScenario: {
+                    traderAssetBalance: BalanceAmountScenario.Higher,
+                    traderAssetAllowance: AllowanceAmountScenario.Higher,
+                    zrxFeeBalance: BalanceAmountScenario.Higher,
+                    zrxFeeAllowance: AllowanceAmountScenario.Higher,
+                },
+                takerStateScenario: {
+                    traderAssetBalance: BalanceAmountScenario.Higher,
+                    traderAssetAllowance: AllowanceAmountScenario.Higher,
+                    zrxFeeBalance: BalanceAmountScenario.Higher,
+                    zrxFeeAllowance: AllowanceAmountScenario.Higher,
+                },
+            };
+            return fillScenario;
+        });
+
+        return fillScenarios;
+    }
+    /**
+     * Recursive implementation of generating all combinations of the supplied
+     * string-containing arrays.
+     */
+    private static _getAllCombinations(arrays: string[][]): string[][] {
```

Comment:
@albrow this is the method I'm using for generating all combinations of scenarios. We will need to genericize it to work for your use case.